Get server time zones

itbw opened the pull request:
Changes in pull request:

* EwsUtilities.PATTERN_SECONDS.
Added escape to dot PATTERN_SECONDS, otherwise it matches any char

* EwsUtilities.getXSDurationToTimeSpan()
Deleted all debug logging, because it does not bring any significant debug information and creates 8 x 108 (number of time zones) messages in debug log

* TimeZoneDefinition.compare(final TimeZoneTransition x, final TimeZoneTransition y) 

  * Added type check before compare. Variable x can only be of type AbsoluteDateTransition (RelativeDayOfMonthTransition and AbsoluteDayOfMonthTransition extend AbsoluteDateTransition). 

  * Variable y can be of type TimeZoneTransition, so added corresponding compare case

* TimeZoneTransition.tryReadElementFromXml()
Refactored getting periods and target groups, because method checks if array does not contain object, and then attempts to get this object from array. Now method correctly gets objects from array, if object does not exist in array an exception will be thrown..

* GetServerTimeZonesRequest
Class and constructor visibility is changed to public, in order to create instance of class in ExchangeService.

* ExchangeService
getServerTimeZones(Iterable<String> timeZoneIds) and getServerTimeZones() are refactored to use GetServerTimeZonesRequest

azurecla commented:
Hi __@itbw__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

azurecla commented:
__@itbw__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

kentongray commented on src/main/java/microsoft/exchange/webservices/data/core/ExchangeService.java:
```diff
@@ -3936,27 +3940,56 @@ public void setExchange2007CompatibilityMode(boolean value) {
    * @return A Collection containing the definitions of the specified time
    * zones.
    */
+	//  public Collection<TimeZoneDefinition> getServerTimeZones(
```
remove commented code

kentongray commented on src/main/java/microsoft/exchange/webservices/data/core/ExchangeService.java:
```diff
@@ -3968,18 +4001,37 @@ public void setExchange2007CompatibilityMode(boolean value) {
-          .inDaylightTime(today), TimeZone.LONG);
-    }
 
+	//  public Collection<TimeZoneDefinition> getServerTimeZones() {
```
same

kentongray commented on src/main/java/microsoft/exchange/webservices/data/core/EwsUtilities.java:
```diff
@@ -871,15 +871,15 @@ public static TimeSpan getXSDurationToTimeSpan(String xsDuration) {
     Matcher m = PATTERN_TIME_SPAN.matcher(xsDuration);
     boolean negative = false;
-    LOG.debug(m.find());
+    //LOG.debug(m.find());
```
should either remove or uncomment

kentongray commented:
@itbw so awesome you took this on this was a pain that i ran deep into and never took the time to figure out what the real problem was (just worked out a way to get what I needed). i just threw on a few comments to help you get this in quicker but it looks good to me, i'm sure @serious6 will have some points to bring up but to help after you clean it up you should rebase and squash the commits into one to help it get merged in quickly

itbw commented:
@kentongray thanks your help! I cleaned the code, rebased and squashed everything in one commit. Hope now merging can be done more quickly.

serious6 commented on src/main/java/microsoft/exchange/webservices/data/core/ExchangeService.java:
```diff
@@ -3928,58 +3932,51 @@ public boolean getExchange2007CompatibilityMode() {
-
+ * @throws Exception 
+   */
+  public Collection<TimeZoneDefinition> getServerTimeZones() throws Exception {
```
any specific reason you are using `Collection` instead of `List`?

itbw replied:
@serious6 The only reason - Collection was used in previous implementation. I didn't want to break API.

serious6 commented on src/main/java/microsoft/exchange/webservices/data/property/complex/time/TimeZoneDefinition.java:
```diff
@@ -114,13 +115,38 @@ public int compare(final TimeZoneTransition x, final TimeZoneTransition y) {
-      final Date secondDateTime = secondTransition.getDateTime();
-
-      return firstDateTime.compareTo(secondDateTime);
+			if (x instanceof AbsoluteDateTransition && y instanceof AbsoluteDateTransition) {
```
some additional unit-tests would be appreciated

itbw replied:
I did't look at the project's unit tests yet, but can write it, if necessary

serious6 replied:
as code-coverage is quite low providing relevant tests is never wrong :wink:

itbw replied:
@serious6 I had a quick look at how to create unit tests and found small problem - all three used class (AbsoluteDateTransition, RelativeDayOfMonthTransition and AbsoluteDayOfMonthTransition) have either protected or default visibility. They can only be instantiated by using TimeZoneTransition.create(...) method, which only create default instance of classes without any values. Values can only be set by using tryReadElementFromXml(...).

My question is following - shall I change visibility of set methods (which does not make much sense by TimeZoneDefenition) or is where any other simple method to create object of required classes with values (mocking EwsServiceXmlReader)? Or is where a general guideline for such cases?

serious6 commented on src/main/java/microsoft/exchange/webservices/data/property/complex/time/TimeZoneTransitionGroup.java:
```diff
@@ -134,7 +134,6 @@ public void writeAttributesToXml(EwsServiceXmlWriter writer)
   public boolean tryReadElementFromXml(EwsServiceXmlReader reader)
       throws Exception {
     reader.ensureCurrentNodeIsStartElement();
-
```
can be reverted?

itbw replied:
Yes, can do

serious6 commented:
@itbw thanks for your contribution. Looking at your commit-msg I am feeling it is a little outdated and might not really describe what you did? Please provide a more relevant message describing why you did these changes.

itbw commented:
@serious6 Good. I'll fix things you pointed and prepare new commit

itbw commented:
@serious6 I refactored TimeZoneDefinition.compare(...) method, because I got better understanding of code after writing unit tests :smile: and added tests to request. Message to pull request is updated.

serious6 commented:
:+1: I think this looks good now

serious6 commented on src/main/java/microsoft/exchange/webservices/data/property/complex/time/TimeZoneDefinition.java:
```diff
@@ -114,13 +115,18 @@ public int compare(final TimeZoneTransition x, final TimeZoneTransition y) {
-      final AbsoluteDateTransition secondTransition = (AbsoluteDateTransition) y;
+      if (x instanceof AbsoluteDateTransition && y instanceof AbsoluteDateTransition) {
+        final AbsoluteDateTransition firstTransition = (AbsoluteDateTransition) x;
+        final AbsoluteDateTransition secondTransition = (AbsoluteDateTransition) y;
```
why not ue firstTransition.compareTo(secondTransition) here?

itbw replied:
Good point. It might make sense to implement compareTo() method for AbsoluteDateTransition, so it can be used in other places later. Plain TimeZoneTransition does not have anything to compare, so we do not really need compareTo() method.

serious6 commented on src/main/java/microsoft/exchange/webservices/data/property/complex/time/TimeZoneDefinition.java:
```diff
@@ -114,13 +115,18 @@ public int compare(final TimeZoneTransition x, final TimeZoneTransition y) {
+        return firstDateTime.compareTo(secondDateTime);
+
+      } else if (y instanceof TimeZoneTransition) {
+        return 1;
```
in case y is instance of TimeZoneTransition x will not be evaluated?

itbw replied:
Yes, each Transitions in TimeZoneDefinition has one plain TimeZoneTransition as a first value. It does not have any DateTime value, so it is always "less" then any instance of AbsoluteDateTransition. In other words, it is the same case as (y == null).

serious6 commented:
@itbw thanks for your contribution!
